### PR TITLE
[PEP 741] gh-107954: Add PyInitConfig C API

### PR DESCRIPTION
**Pull request by vstinner**
Add PyInitConfig functions:

* PyInitConfig_Python_New()
* PyInitConfig_Isolated_New()
* PyInitConfig_Free(config)
* PyInitConfig_SetInt(config, key, value)
* PyInitConfig_SetStr(config, key, value)
* PyInitConfig_SetWStr(config, key, value)
* PyInitConfig_SetStrList(config, key, length, items)
* PyInitConfig_SetWStrList(config, key, length, items)
* PyInitConfig_GetErrorMsg(config)

Add also functions using it:

* Py_InitializeFromInitConfig(config)
* Py_ExitWithInitConfig(config)

Changes:

* Add Include/initconfig.h header.

**Discussion:** https://discuss.python.org/t/fr-allow-private-runtime-config-to-enable-extending-without-breaking-the-pyconfig-abi/18004/20

* Issue: gh-107954

### CODE REVIEW

**Comment by vstinner** [comment]
Discussion: https://discuss.python.org/t/fr-allow-private-runtime-config-to-enable-extending-without-breaking-the-pyconfig-abi/18004/20

**Comment by vstinner** [comment]
TODO: once we reach an consensus on the API, I will document the API.

**Comment by vstinner** [comment]
I documented the API. I decided to add a new `Doc/c-api/config.rst` documentation for that.

File: Doc/c-api/config.rst [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+.. c:function:: int PyInitConfig_SetInt(PyInitConfig *config, const char *key, int64_t value)
+
+   Set an integer configuration option.
+
+   Return 0 on success, or return -1 on error.
+
+
+.. c:function:: int PyInitConfig_SetStr(PyInitConfig *config, const char *key, const char *value)
+
+   Set a string configuration option from a bytes string.
```
Comment (encukou) [inline Doc/c-api/config.rst]: I'd prefer to be explicit in new docs:
```suggestion
   Set a string configuration option from a NUL-terminated bytes string.
```

File: Doc/c-api/config.rst [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+   The bytes string is decoded by Py_DecodeLocale(). Preinitialize Python if
+   needed to ensure that encodings are properly configured.
+
+   Return 0 on success, or return -1 on error.
```
Comment (encukou) [inline Doc/c-api/config.rst]: And I'd make it more obvious that that the error is stored in the `PyInitConfig`:
```suggestion
   Return 0 on success.
   On error, return -1. The error can be retrieved with
   :c:expr:`PyInitConfig_GetErrorMsg(config)`
```

File: Doc/c-api/config.rst [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+.. c:function:: int PyInitConfig_SetWStr(PyInitConfig *config, const char *key, const wchar_t *value)
+
+   Set a string configuration option from a wide string.
+
+   Preinitialize Python if needed.
```
Comment (encukou) [inline Doc/c-api/config.rst]: It's not clear if the function preinitializes Python, or if the *user* needs to do it if needed.
```suggestion
   If Python is not yet preinitialized, this function preinitializes it.
```

File: Doc/c-api/config.rst [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+
+.. c:function:: int Py_InitializeFromInitConfig(PyInitConfig *config)
+
+   Initialize Python from the initialization configuration.
+
+   Return ``0`` on success.
+
+   Return ``-1`` if Python wants to exit and on error. Call
+   :c:func:`Py_ExitWithInitConfig` in this case.
```
Comment (encukou) [inline Doc/c-api/config.rst]: What happens if you don't call it? (If it needs to be called, why does `Py_InitializeFromInitConfig` not call it itself?)

File: Doc/c-api/config.rst [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+   Similar to :c:func:`PyInitConfig_Python_New`, but use :ref:`Isolated
+   Configuration <init-isolated-conf>` default values.
+
+
+.. c:function:: void PyInitConfig_Free(PyInitConfig *config)
+
+   Free memory of a initialization configuration.
+
+
+.. c:function:: int PyInitConfig_SetInt(PyInitConfig *config, const char *key, int64_t value)
```
Comment (encukou) [inline Doc/c-api/config.rst]: It's not clear that `key` should be a name of a `PyConfig` member. (Should it?)

File: Doc/c-api/config.rst [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+   Set a string configuration option from a wide string.
+
+   Preinitialize Python if needed.
+
+   Return ``0`` on success, or return ``-1`` on error.
+
+
+.. c:function:: int PyInitConfig_SetStrList(PyInitConfig *config, const char *key, size_t length, char * const *items)
+
+   Set a string list configuration option from bytes strings.
```
Comment (encukou) [inline Doc/c-api/config.rst]:
```suggestion
   Set a string list configuration option from an array of NUL-terminated bytes strings.
```

**Review by encukou — COMMENTED** [review]
Thank you!
Now that I see all the docs, it looks the design is missing a function: https://discuss.python.org/t/fr-allow-private-runtime-config-to-enable-extending-without-breaking-the-pyconfig-abi/18004/49?u=encukou

I also have a few docs nitpicks. I didn't look very closely at the implementation yet.

File: Doc/c-api/config.rst (hunk 2 above)
Comment (vstinner) [reply Doc/c-api/config.rst]: I updated the doc.

File: Doc/c-api/config.rst (hunk 4 above)
Comment (vstinner) [reply Doc/c-api/config.rst]: The idea is to let the caller decide how to handle an error or an exit code. When you embed Python, you may not want the process to exit. It's a similar design than `Py_InitializeFromConfig()` and the `PyStatus` C API.

> What happens if you don't call it?

Python is in an undefined state. The error/exit code stays in PyInitConfig. It's safer to handle the init error :-)

**Comment by vstinner** [comment]
This change is quite big, so I moved PyConfig_Get() API to a separated PR: PR #112609.

**Comment by vstinner** [comment]
This API is kind of complicated (there are multiple use cases). I created PR #112609 as a starting point, just to add two functions. But apparently, adding PyConfig_Get() and PyConfig_GetInt() functions is more complicated than what I expected. I don't have the bandwidth to deal with such large API for now, so I prefer to close the PR. See also my [comment on the PyConfig_Get() PR](https://github.com/python/cpython/pull/112609#issuecomment-1864287222).

**Comment by vstinner** [comment]
New PR: https://github.com/python/cpython/pull/123502

**Pull request closed without merging**